### PR TITLE
9 mining middleware flow

### LINKED ISSUE #9: Transactions are picked from the queue and added to chain
**Description:**

When the mining-API pops a transaction from its transaction queue, it should mine a new block for that transaction and send it to the blockchain to be added to the chain. Mining will be moved to the client-side, with the mining-API orchestrating the mining session, in Sprint 3.

**Acceptance Criteria:**
- [x] Transaction is popped off the queue
- [x] Proof is found for that transaction
- [x] The proof and transaction data are sent to the blockchain
- [x] The blockchain creates a new block with the data sent from the mining-API
- [x] The block is added successfully

### PR DESCRIPTION
**Pull request by abhisachdev17**
## Summary of Changes
- Added mining flow support to the mining middleware
- blockchain now supports the mining process with validity logic being implemented in mining
- Updated and wrote new tests

# TODO
- [x] Tests to validate entire socket flow

### CODE REVIEW

File: src/backend/blockchain/src/routes.py [hunk 1]
```diff
@@ -25,30 +25,39 @@ def get_chain():
...
+                    new_transaction,
+                    time.time(),
+                    nonce,
+                    hash,
+                    blockchain.get_last_block().hash,
+                    miner_id, 
+                    Blockchain.COINBASE_AMOUNT
+                )
+
+     # We guarantee that all the necessary validation has been done by this point, so simply add the
```
Comment (ZackHolmberg) [inline src/backend/blockchain/src/routes.py]: Comment indentation a bit off

File: src/backend/services/mining-api/endpoints.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  "minerId": string
+}
+```
+
+> broadcast to all clients broadcasts "findProof"
+
+Sends JSON:
+
+```json
+Transaction Object
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/endpoints.md:62]: Can you please change this to be the explicit object, I know we use it many other places but always best to be explicit for 3rd party devs

File: src/backend/services/mining-api/src/routes.py [hunk 3]
```diff
@@ -31,12 +35,19 @@ def mine(transaction):
 
 def send_to_connected_clients(transaction):
     # ToDo
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: I assume this comment can be removed now

File: src/backend/services/mining-api/src/routes.py [hunk 4]
```diff
@@ -54,20 +65,54 @@ def add_data_to_queue():
...
+    valid = (hash_.startswith('0' * difficulty))
+    return valid and hash_ == sha256(
+        (str(nonce) + ongoing_transaction["to"] + ongoing_transaction["from"] + str(ongoing_transaction["amount"]) + ongoing_transaction["id"] + ongoing_transaction["signature"]).encode('utf-8')
+      ).hexdigest();
+
+@socketio.on('connect')
+def client_connect():
+    global connected_clients
+    connected_clients += 1
+    print(connected_clients)
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Should remove this print and the print on line 85 before the merge

File: src/backend/services/mining-api/src/routes.py [hunk 5]
```diff
@@ -54,20 +65,54 @@ def add_data_to_queue():
...
 def handle_proofs(message):
-    transactions.ready_to_mine()
-    # TODO if proof for wrong id is received compare it with the current ongoing proof
-
-    # TODO send the transaction to blockchain
-
-    # TODO reward the miner (create a new transaction and send it to blockchain)
-
-    socketio.emit("stop_finding", message)
+    global ongoing_proof, ongoing_transaction, COINBASE_AMOUNT, blockchain_url, blockchain_wallet_url, transactions
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Can remove COINBASE_AMOUNT

File: src/backend/services/mining-api/src/routes.py [hunk 6]
```diff
@@ -54,20 +65,54 @@ def add_data_to_queue():
...
+    
+    if (ongoing_proof == message['id']):
+        if valid_proof(message["proof"], message["nonce"]):
+
+            # ignore the rest of the clients that try to send the proof
+            ongoing_proof = None
+            ongoing_transaction = None
+            
+            socketio.emit("stopProof", {})
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Can you leave the 2nd arg empty or as None? Don't think we need to send an empty object. Same on line 115

File: src/backend/services/mining-api/src/routes.py [hunk 7]
```diff
@@ -54,20 +65,54 @@ def add_data_to_queue():
...
+            # new transactions are now ready to be mined    
+            transactions.ready_to_mine()
+
+            block_data = {}
+            # send transactions to blockchain
+            for key, value in message.items():
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py:99]: Not a suggestion but interesting that the message from the socket comes in the form of a dict and not an object! Wouldn't have thought that

File: src/backend/services/mining-api/tests/functional/test_routes.py [hunk 8]
```diff
@@ -22,6 +33,21 @@ def test_home_page(test_client):
     assert response.status_code == 200
     assert b"Hello From the Mining" in response.data
 
+def test_queueing(test_client, json_header):
+    # test GET query on '/' route
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/tests/functional/test_routes.py]: This comment and line 46 can be removed

**Review by ZackHolmberg — APPROVED** [review]
Looks pretty great to me, a couple of small changes but nothing functional so approved!

File: src/backend/blockchain/src/blockchain.py [hunk 9]
```diff
@@ -3,9 +3,11 @@
 from .transaction import Transaction
 from .exceptions import WalletException
 
+
 class Blockchain:
 
     difficulty = 4
+    COINBASE_AMOUNT = 10
```
Comment (madisonfines) [inline src/backend/blockchain/src/blockchain.py:10]: Is this coinbase_amount used anywhere?

File: src/backend/blockchain/src/routes.py [hunk 10]
```diff
@@ -25,30 +25,39 @@ def get_chain():
...
     try :
         new_transaction = Transaction(
-            data["from"], 
-            data["to"], 
+            data["from"],
+            data["to"],
             data["amount"]
         )
+        miner_id = data["minerId"]
+        _hash = data["proof"]
```
Comment (madisonfines) [inline src/backend/blockchain/src/routes.py:38]: Should this be _hash or hash?

**Review by madisonfines — APPROVED** [review]

File: src/backend/blockchain/src/blockchain.py (hunk 9 above)
Comment (ZackHolmberg) [reply src/backend/blockchain/src/blockchain.py:10]: It's used on line 52 in `blockchain/src/routes.py`

File: src/backend/services/mining-api/endpoints.md (hunk 2 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/endpoints.md:62]: I defined the object at the beginning of the file

File: src/backend/services/mining-api/src/routes.py (hunk 6 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/src/routes.py]: I am not sure if that will play well with socketio, I will test it out

File: src/backend/blockchain/src/routes.py (hunk 10 above)
Comment (abhisachdev17) [reply src/backend/blockchain/src/routes.py:38]: hash seems to a python protected variable so changed it to _hash instead

File: src/backend/services/mining-api/tests/functional/test_routes.py (hunk 8 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/tests/functional/test_routes.py]: removed

File: src/backend/services/mining-api/src/routes.py (hunk 7 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/src/routes.py:99]: Yeah looks like socketio handles JSON to dictionary conversion, which is pretty interesting.

File: src/backend/services/mining-api/src/routes.py (hunk 6 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/src/routes.py]: It works! javascript will receive a null value (`undefined`)

File: src/backend/services/mining-api/src/routes.py (hunk 5 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/src/routes.py]: done!

File: src/backend/blockchain/src/routes.py (hunk 10 above)
Comment (abhisachdev17) [reply src/backend/blockchain/src/routes.py:38]: clarification: hash() is an inbuilt method in python so cannot use that name.

File: src/backend/services/mining-api/endpoints.md (hunk 2 above)
Comment (ZackHolmberg) [reply src/backend/services/mining-api/endpoints.md:62]: Ah yes I didnt see that before, thanks for pointing that out

File: src/backend/services/mining-api/src/routes.py [hunk 11]
```diff
@@ -13,30 +17,38 @@
...
 
-COINBASE_AMOUNT = 10
-
 blockchain_url = BisonCoinUrls.blockchain_url
 
+ongoing_proof = None
+ongoing_transaction = None
+connected_clients = 0
+
 # This is temporary till we move mining to user devices
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Can remove this comment

File: src/backend/services/mining-api/src/routes.py [hunk 12]
```diff
@@ -13,30 +17,38 @@
...
 
-# transactions = MiningPool(sendToConnectedClients, True)
-transactions = MiningPool(mine, True)
+    while (connected_clients == 0) : 
+        sleep(0.5)
+
+    socketio.emit("findProof", transaction)
+
+transactions = MiningPool(send_to_connected_clients, True)
+# transactions = MiningPool(mine, True)
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: Remove comment

**Review by ZackHolmberg — APPROVED** [review]

File: src/backend/services/mining-api/src/routes.py (hunk 12 above)
Comment (abhisachdev17) [reply src/backend/services/mining-api/src/routes.py]: done!

**Pull request merged**